pkg/applicationserver/io/pubsub/provider/mqtt/driver.go

```diff
@@ -0,0 +1,280 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mqtt
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+type topic struct {
+	client  mqtt.Client
+	topic   string
+	timeout time.Duration
+	qos     byte
+}
+
+var errNilClient = errors.DefineInvalidArgument("nil_client", "client is nil")
+
+// OpenTopic returns a *pubsub.Topic that publishes to the given topic name with the given MQTT client.
+func OpenTopic(client mqtt.Client, topicName string, timeout time.Duration, qos byte) (*pubsub.Topic, error) {
+	dt, err := openDriverTopic(client, topicName, timeout, qos)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(dt, nil), nil
+}
+
+func openDriverTopic(client mqtt.Client, topicName string, timeout time.Duration, qos byte) (driver.Topic, error) {
+	if client == nil {
+		return nil, errNilClient
```

Comment:
This is required for [the test harness](https://github.com/google/go-cloud/blob/17853d2a51cb6d8ce9e825872467db3ced8d3210/pubsub/drivertest/drivertest.go#L49) - there is no way to have a topic that does not exist in MQTT, and as such we mimic it using the nil topic. You are very much correct that this should never happen otherwise.